File: 1 C/2_pset/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long cardNum = get_long("Credit card's number: ");
    long cardTemp = cardNum;
    int cardCalc = 0, sum1 = 0, sum2 = 0;

    //Looping while the credit card number is greater than 0 to get sum1.
    while (cardTemp > 0)
    {
        //Using cardCalc to sum up every other number, starting with the number’s second-to-last digit. Note that cardCalc is an int.
        cardCalc = ((cardTemp / 10) % 10) * 2;
        //Store the sum of the numbers in sum1.
        if (cardCalc < 10)
        {
            sum1 += cardCalc;
        }
        //In case the number is greater than 10, because we need those products digits and numbers won't exceed 19,
        //we can substract 10 from the total and add 1 for the second digit.
        else
        {
            sum1 += (cardCalc - 10) + 1;
        }
        //Now that we have the numbers stored in sum1, we proceed to:
        //Divide cardNum so we move the decimal to the next number needed.
        cardTemp /= 100;
    }
    cardTemp = cardNum;
    cardCalc = 0;

    //Same as in sum1 but with sum2, the other digits
    while (cardTemp > 0)
    {
        cardCalc = cardTemp % 10;
        sum2 += cardCalc;
        cardTemp /= 100;
    }

    cardCalc = 0;
    cardCalc = (sum1 + sum2) % 10;

    //Now we check if the checksum equals 0, if so we proceed to evaluate whether is an AMEX, MASTERCARD or VISA.
    if (cardCalc != 0)
    {
        printf("INVALID\n");
        return 0;
    }
    else if (cardCalc == 0)
    {
        if ((cardNum / 10000000000000) == 37 || (cardNum / 10000000000000) == 34)
        {
            printf("AMEX\n");
            return 0;
        }
        if ((cardNum / 100000000000000) > 50 && (cardNum / 100000000000000) < 56)
        {
            printf("MASTERCARD\n");
            return 0;
        }
        if ((cardNum / 1000000000000) == 4 || (cardNum / 1000000000000000) == 4)
        {
            printf("VISA\n");
            return 0;
        }
        else
        {
            printf("INVALID\n");
            return 0;
        }
    }
}

File: 6 Python/2_pset/sentimental-credit/credit.py
# TODO
import sys

credit = input("Number: ")
sum1 = []
sum2 = 0
total = 0
if len(credit) > 16:
    print("INVALID")
    sys.exit
elif len(credit) == 16:
    for i in range(1, len(credit), 2):
        sum2 += int(credit[i])

    for i in range(0, len(credit), 2):
        sum1.append(int(credit[i]) * 2)
else:
    for i in range(0, len(credit), 2):
        sum2 += int(credit[i])

    for i in range(1, len(credit), 2):
        sum1.append(int(credit[i]) * 2)

for i in sum1:
    if i >= 10:
        i = (i % 10) + 1
    total += i

total = total + sum2

if total % 10 == 0:
    total = int(str(credit)[:2])
    total2 = int(str(credit)[:1])
    if total2 == 4:
        print("VISA")
        sys.exit
    match total:
        case 34 | 37:
            print("AMEX")
        case 51 | 52 | 53 | 54 | 55:
            print("MASTERCARD")
        case _:
            print("INVALID")

else:
    print("INVALID")